File: Zeze/Transaction/Procedure.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Zeze.Transaction
{
    public class Procedure
    {
        public const long Success = 0;
        public const long Exception = -1;
        public const long TooManyTry = -2;
        public const long NotImplement = -3;
        public const long Unknown = -4;
        public const long ErrorSavepoint = -5;
        public const long LogicError = -6;
        public const long RedoAndRelease = -7;
        public const long AbortException = -8;
        public const long ProviderNotExist = -9;
        public const long Timeout = -10;
        public const long CancelException = -11;
        public const long DuplicateRequest = -12;
        public const long ErrorRequestId = -13;
        public const long ErrorSendFail = -14;
        public const long RaftRetry = -15;
        public const long RaftApplied = -16;
        public const long RaftExpired = -17;
        // >0 用户自定义。

        private static readonly NLog.Logger logger = NLog.LogManager.GetCurrentClassLogger();

        public Application Zeze { get; }

        public Func<Task<long>> Action { get; set; }

        public string ActionName { get; set; } // 用来统计或者日志
        public TransactionLevel TransactionLevel { get;}

        // 用于继承方式实现 Procedure。
        public Procedure(Application app)
        {
            Zeze = app;
        }

        public object UserState { get; set; }

        public Zeze.Net.Protocol Rpc { get; set; }

        public Procedure(Application app, Func<Task<long>> action, string actionName, TransactionLevel level, object userState)
        {
            Zeze = app;
            Action = action;
            ActionName = actionName;
            TransactionLevel = level;
            UserState = userState;
            if (null == UserState) // 没指定，就从当前存储过程继承。嵌套时发生。
                UserState = Transaction.Current?.TopProcedure?.UserState;
        }

        public static volatile Action<Exception, long, Procedure, string> LogAction = DefaultLogAction;

        public static void DefaultLogAction(Exception ex, long result, Procedure p, string message)
        {
            NLog.LogLevel ll = (null != ex) ? NLog.LogLevel.Error
                : (0 != result) ? p.Zeze.Config.ProcessReturnErrorLogLevel
                : NLog.LogLevel.Trace;

            var module = "";
            if (result > 0)
                module = "@" + IModule.GetModuleId(result) + ":" + IModule.GetErrorCode(result);

            logger.Log(ll, ex, $"Procedure={p} Return={result}{module} {message} UserState={p.UserState}");
        }

        public void Execute(Net.Protocol from = null, Action<Net.Protocol, long> actionWhenError = null)
        {
            ExecutionContext.SuppressFlow();
            Task.Run(async () => await Util.Mission.CallAsync(this, from, actionWhenError));
            ExecutionContext.RestoreFlow();
        }

        /// <summary>
        /// 创建 Savepoint 并执行。
        /// 嵌套 Procedure 实现，
        /// </summary>
        /// <returns></returns>
        public async Task<long> CallAsync()
        {
            if (null == Transaction.Current)
            {
                try
                {
                    // 有点奇怪，Perform 里面又会回调这个方法。这是为了把主要流程都写到 Transaction 中。
                    return await Transaction.Create(Zeze.Locks).Perform(this);
                }
                finally
                {
                    // Transaction.AsyncLocal 到达这里肯定是null，作为概念，执行一次Destroy。
                    Transaction.Destroy();
                }
            }

            Transaction currentT = Transaction.Current;
            currentT.Begin();
            currentT.ProcedureStack.Add(this);

            try
            {
                long result = await Process();
                currentT.VerifyRunning(); // 防止应用抓住了异常，通过return方式返回。

                if (Success == result)
                {
                    currentT.Commit();
#if ENABLE_STATISTICS
                    ProcedureStatistics.Instance.GetOrAdd(ActionName).GetOrAdd(result).IncrementAndGet();
#endif
                    return Success;
                }
                currentT.Rollback();
                LogAction?.Invoke(null, result, this, "");
#if ENABLE_STATISTICS
                ProcedureStatistics.Instance.GetOrAdd(ActionName).GetOrAdd(result).IncrementAndGet();
#endif
                return result;
            }
            catch (GoBackZezeException gobackzeze)
            {
                // 单独抓住这个异常，是为了能原样抛出，并且使用不同的级别记录日志。
                // 对状态正确性没有影响。
                currentT.Rollback();
                logger.Debug(gobackzeze);
                throw gobackzeze;
            }
            catch (Exception e)
            {
                currentT.Rollback();
                LogAction?.Invoke(e, Exception, this, "");
#if ENABLE_STATISTICS
                ProcedureStatistics.Instance.GetOrAdd(ActionName).GetOrAdd(Exception).IncrementAndGet();
#endif
                currentT.VerifyRunning();
#if DEBUG
                // 对于 unit test 的异常特殊处理，与unit test框架能搭配工作
                if (e.GetType().Name == "AssertFailedException")
                {
                    throw;
                }
#endif
                return e is TaskCanceledException ? CancelException : Exception; // 回滚当前存储过程，不中断事务，外层存储过程判断结果自己决定是否继续。
            }
            finally
            {
                currentT.ProcedureStack.RemoveAt(currentT.ProcedureStack.Count - 1);
            }
        }

        protected virtual async Task<long> Process()
        {
            if (null != Action)
                return await Action();
            return NotImplement;
        }

        public long CallSynchronously()
        {
            var atask = CallAsync();
            atask.Wait();
            return atask.Result;
        }

        public override string ToString()
        {
            // GetType().FullName 仅在用继承的方式实现 Procedure 才有意义。
            return (null != Action) ? ActionName : this.GetType().FullName;
        }
    }
}

File: ZezeJava/src/Zeze/Transaction/Procedure.java
package Zeze.Transaction;

import java.util.concurrent.Callable;

import Zeze.Util.TaskCanceledException;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import Zeze.*;

public class Procedure {
	public static final long Success = 0;
	public static final long Excption = -1;
	public static final long TooManyTry = -2;
	public static final long NotImplement = -3;
	public static final long Unknown = -4;
	public static final long ErrorSavepoint = -5;
	public static final long LogicError = -6;
	public static final long RedoAndRelease = -7;
	public static final long AbortException = -8;
	public static final long ProviderNotExist = -9;
	public static final long Timeout = -10;
	public static final long CancelExcption = -11;
	public static final long DuplicateRequest = -12;
	public static final long ErrorRequestId = -13;
	// >0 用户自定义。

	private static final Logger logger = LogManager.getLogger(Procedure.class);

	private Application Zeze;
	public final Application getZeze() {
		return Zeze;
	}

	private Callable<Long> Action;
	public final Callable<Long> getAction() {
		return Action;
	}
	public final void setAction(Callable<Long> value) {
		Action = value;
	}

	private String ActionName;
	public final String getActionName() {
		return ActionName;
	}
	public final void setActionName(String value) {
		ActionName = value;
	}

	// 用于继承方式实现 Procedure。
	public Procedure(Application app) {
		Zeze = app;
	}

	private Object UserState;
	public final Object getUserState() {
		return UserState;
	}
	public final void setUserState(Object value) {
		UserState = value;
	}

	public Procedure(Application app, Callable<Long> action, String actionName, Object userState) {
		Zeze = app;
		setAction(action);
		setActionName(actionName);
		setUserState(userState);
		if (null == getUserState()) { // 没指定，就从当前存储过程继承。嵌套时发生。
			setUserState(Transaction.getCurrent() == null
					? null : (Transaction.getCurrent().getTopProcedure() == null
					? null : Transaction.getCurrent().getTopProcedure().getUserState()));
		}
	}

	/** 
	 创建 Savepoint 并执行。
	 嵌套 Procedure 实现，
	 
	 @return 
	*/
	public final long Call() {
		if (null == Transaction.getCurrent()) {
			try {
				// 有点奇怪，Perform 里面又会回调这个方法。这是为了把主要流程都写到 Transaction 中。
				return Transaction.Create(Zeze.getLocks()).Perform(this);
			}
			finally {
				Transaction.Destroy();
			}
		}

		Transaction currentT = Transaction.getCurrent();
		currentT.Begin();
		currentT.getProcedureStack().add(this);

		try {
			var result = Process();
			if (Success == result) {
				currentT.Commit();

				ProcedureStatistics.getInstance().GetOrAdd(getActionName()).GetOrAdd(result).incrementAndGet();

				return Success;
			}
			currentT.Rollback();

			var module = "";
			if (result > 0) {
				module = "@" + IModule.GetModuleId(result) + ":" + IModule.GetErrorCode(result);
			}
			logger.log(getZeze().getConfig().getProcessReturnErrorLogLevel(),
					"Procedure {} Return{}@{} UserState={}", this, result, module, getUserState());

			ProcedureStatistics.getInstance().GetOrAdd(getActionName()).GetOrAdd(result).incrementAndGet();

			return result;
		}
		catch (RedoException e) {
			currentT.Rollback();
			throw e;
		}
		catch (AbortException e) {
			currentT.Rollback();
			throw e;
			// 抛出这个异常，中断事务，跳过所有嵌套过程直到最外面。 e;
		}
		catch (RedoAndReleaseLockException e2) {
			currentT.Rollback();

			ProcedureStatistics.getInstance().GetOrAdd(getActionName()).GetOrAdd(RedoAndRelease).incrementAndGet();

			throw e2; // 抛出这个异常，打断事务，跳过所有嵌套过程直到最外面。会重做。
		}
		catch (TaskCanceledException e3) {
			currentT.Rollback();
			logger.error("Procedure {} Exception UserState={}", this, getUserState(), e3);

			ProcedureStatistics.getInstance().GetOrAdd(getActionName()).GetOrAdd(Excption).incrementAndGet();

			return CancelExcption; // 回滚当前存储过程，不中断事务，外层存储过程判断结果自己决定是否继续。
		}
		catch (Throwable e4) {
			currentT.Rollback();
			logger.error("Procedure {} Exception UserState={}", this, getUserState(), e4);

			ProcedureStatistics.getInstance().GetOrAdd(getActionName()).GetOrAdd(Excption).incrementAndGet();

			// 对于 unit test 的异常特殊处理，与unit test框架能搭配工作
			if (e4.getClass().getSimpleName().equals("AssertFailedException")) {
				throw e4;
			}

			return Excption; // 回滚当前存储过程，不中断事务，外层存储过程判断结果自己决定是否继续。
		}
		finally {
			currentT.getProcedureStack().remove(currentT.getProcedureStack().size() - 1);
		}
	}

	protected long Process() {
		if (null != Action) {
			try {
				return Action.call();
			} catch (Throwable e) {
				throw new RuntimeException(e);
			}
		}
		return NotImplement;
	}

	@Override
	public String toString() {
		// GetType().FullName 仅在用继承的方式实现 Procedure 才有意义。
		return (null != getAction()) ? getActionName() : this.getClass().getName();
	}
}